src/lua/init.lua

```diff
@@ -86,81 +106,81 @@ local function load_lua(file)
+    paths = type(paths) == 'table' and paths or { paths }
+    templates = templates or {}
 
-local function search_cwd_lua(name)
-    local path = "./?.lua;./?/init.lua"
-    return package.searchpath(name, path)
-end
+    if #templates == 0 then
+        return paths
+    end
+
+    if #paths == 0 then
+        return templates
+    end
 
-local function traverse_rocks(name, pathes_search)
-    local cwd = fio.cwd()
-    local index = string.len(cwd) + 1
-    local strerr = ""
-    while index ~= nil do
-        cwd = string.sub(cwd, 1, index - 1)
-        for i, path in ipairs(pathes_search) do
-            local file, err = package.searchpath(name, cwd .. path)
-            if err == nil then
-                return file
-            end
-            strerr = strerr .. err
+    for _, path in ipairs(paths) do
+        for _, template in pairs(templates) do
+            table.insert(searchpaths, fio.pathjoin(path, template))
         end
-        index = string.find(cwd, "/[^/]*$")
     end
-    return nil, strerr
+
+    return searchpaths
 end
 
-local function search_rocks_lua(name)
-    local pathes_search = {
-        "/.rocks/share/tarantool/?.lua;",
-        "/.rocks/share/tarantool/?/init.lua;",
-    }
-    return traverse_rocks(name, pathes_search)
+local LIB_TEMPLATES = { '?.'..soext }
+local LUA_TEMPLATES = { '?.lua', '?/init.lua' }
+local ROCKS_LIB_TEMPLATES = { ROCKS_LIB_PATH .. '/?.'..soext }
+local ROCKS_LUA_TEMPLATES = { ROCKS_LUA_PATH .. '/?.lua', ROCKS_LUA_PATH .. '/?/init.lua' }
+
+local function search_lib(name, path)
+    return package.searchpath(name, table.concat(merge_strings(path, LIB_TEMPLATES), ';'))
 end
 
-local function search_rocks_lib(name)
-    local pathes_search = {
-        "/.rocks/lib/tarantool/?."..soext
-    }
-    return traverse_rocks(name, pathes_search)
+local function search_lua(name, path)
+    return package.searchpath(name, table.concat(merge_strings(path, LUA_TEMPLATES), ';'))
+end
+
+local function traverse_path(path)
+    path = path:sub(1, 1) == '/' and path or fio.abspath(path)
+    local paths = { path }
+
+    while path ~= '/' do
+        path = fio.dirname(path)
+        table.insert(paths, path)
+    end
+
+    return paths
 end
 
-local function cwd_loader_func(lib)
-    local search_cwd = lib and search_cwd_lib or search_cwd_lua
-    local load_func = lib and load_lib or load_lua
+local function get_search_func(opts)
+    local path_fn = opts.path_fn
+    if type(opts.path_fn) ~= 'function' then
+        path_fn = function() return opts.path_fn end
+    end
+
     return function(name)
-        if not name then
-            return "empty name of module"
-        end
-        local file, err = search_cwd(name)
-        if not file then
-            return err
-        end
-        local loaded, err = load_func(file, name)
-        if err == nil then
-            return loaded
-        else
-            return err
-        end
+        local path = path_fn() or '.'
```

Comment:
`or '.'` is needed only to use cwd when approot is not set? Why not to skip this loader in the case at all? Otherwise you'll search in cwd twice (for regular modules, not rocks).

Reply:
How do you skip a loader in Lua?

Reply:
Return empty string.

Anyway, we discussed this voicely with you and decided to 'merge' cwd and appdir loaders (use appdir when available, otherwise cwd), so skipping is not needed.

Reply:
Removed possible duplication of loaders, but let `path_fn() or '.'` stay for the say of consistency

Reply:
Now searchroot can return `nil`, so `or '.'` is needed. This however the only case when we pass a relative path into `traverse_path()`, so I would use `or fio.cwd()` here. We however should leave `fio.abspath()` in `traverse_path()` to suppress `/../` parts.

Reply:
Ouch, no, `/../` parts are suppressed in `setsearchroot()`.

Reply:
Now the only style question `or '.'` vs `or fio.cwd()` is actual here. It is not so important. Marked as resolved.